Sources/Core/Client/WebSocket/WebSocket.swift

```diff
@@ -161,11 +177,87 @@ public final class WebSocket {
         }
     }
     
-    func clearStateAfterDisconnect() {
+    private func isTokenExpired() -> Bool {
+        guard let lastJSONError = lastJSONError else {
+            return false
+        }
+        
+        let isTokenExpired = lastJSONError.code == ClientErrorResponse.tokenExpiredErrorCode
+        return isTokenExpired && Client.shared.touchTokenProvider()
+    }
+    
+    private func disconnectedNoInternet() {
+        logger?.log("💔🕸 Disconnected: No Internet")
+        clearStateAfterDisconnect()
+        consecutiveFailures = 0
+    }
+    
+    private func disconnected(_ error: Error? = nil) -> Connection {
+        logger?.log("💔🤔 Disconnected")
+        clearStateAfterDisconnect()
+        
+        if let error = error {
+            var errorMessage = "🦄💔😡 Disconnected by error"
+            
+            if let lastJSONError = lastJSONError {
+                errorMessage += ": \(lastJSONError)"
+            }
+            
+            logger?.log(error, message: errorMessage)
+            ClientLogger.showConnectionAlert(error, jsonError: lastJSONError)
+            
+            if !willReconnectAfterError(error) {
+                consecutiveFailures = 0
+                
+                if let lastJSONError = lastJSONError, isStopError(error) {
+                    return .disconnected(lastJSONError)
+                }
+            }
+        } else {
+            consecutiveFailures = 0
+        }
+        
+        return .notConnected
+    }
+    
+    private func willReconnectAfterError(_ error: Swift.Error) -> Bool {
+        if isStopError(error) {
+            return false
+        }
+        
+        if InternetConnection.shared.isAvailable {
+            reconnect()
```

Comment:
`willReconnectAfterError` calling `reconnect` is unexpected, to me it doesn't sound like a mutating func

Reply:
I renamed it.